File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/BaseViewModel.java
package com.websarva.wings.android.zuboradiary.ui;

import android.os.Looper;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.websarva.wings.android.zuboradiary.data.AppMessage;
import com.websarva.wings.android.zuboradiary.data.AppMessageList;

import java.util.Objects;

public abstract class BaseViewModel extends ViewModel {

    private final MutableLiveData<AppMessageList> appMessageBufferList = new MutableLiveData<>();

    public BaseViewModel() {
        initializeAppMessageList();
    }

    protected void initializeAppMessageList() {
        appMessageBufferList.setValue(new AppMessageList());
    }

    /**
     * 本メソッドは継承先のクラス毎に処理するタイミングが異なるので、Override後、継承クラスのコンストラクタに含めること。
     * */
    protected abstract void initialize();

    protected final void addAppMessage(AppMessage appMessage) {
        Objects.requireNonNull(appMessage);

        AppMessageList currentList = appMessageBufferList.getValue();
        Objects.requireNonNull(currentList);
        AppMessageList updateList = currentList.add(appMessage);

        boolean isMainThread = (Looper.getMainLooper().getThread() == Thread.currentThread());
        if (isMainThread) {
            appMessageBufferList.setValue(updateList);
        } else {
            appMessageBufferList.postValue(updateList);
        }
    }

    public final void triggerAppMessageBufferListObserver() {
        AppMessageList currentList = appMessageBufferList.getValue();
        appMessageBufferList.setValue(new AppMessageList());
        appMessageBufferList.setValue(currentList);
    }

    public final void removeAppMessageBufferListFirstItem() {
        AppMessageList currentList = appMessageBufferList.getValue();
        Objects.requireNonNull(currentList);
        AppMessageList updateList = currentList.removeFirstItem();
        appMessageBufferList.setValue(updateList);
    }

    protected final boolean equalLastAppMessage(AppMessage appMessage) {
        Objects.requireNonNull(appMessage);

        AppMessageList currentList = appMessageBufferList.getValue();
        Objects.requireNonNull(currentList);
        return currentList.equalLastItem(appMessage);
    }

    public final LiveData<AppMessageList> getAppMessageBufferListLiveData() {
        return appMessageBufferList;
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/BaseViewModel.kt
package com.websarva.wings.android.zuboradiary.ui

import android.util.Log
import androidx.lifecycle.ViewModel
import com.websarva.wings.android.zuboradiary.data.AppMessage
import com.websarva.wings.android.zuboradiary.data.AppMessageList
import com.websarva.wings.android.zuboradiary.getLogTag
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow

abstract class BaseViewModel : ViewModel() {

    private val logTag = getLogTag()

    // MEMO:ViewModelのスコープ範囲がActivityの時に各プロパティを初期化できるように抽象メソッドinitialize()を用意しているが、
    //      appMessageBufferListに関しては主にエラー表示となるため、ViewModelの初期化時のみの初期化する。
    private val _appMessageBufferList = MutableStateFlow(AppMessageList())
    val appMessageBufferList
        get() = _appMessageBufferList.asStateFlow()

    open fun initialize() {
        Log.d(logTag, "addAppMessage()")
    }

    protected fun addAppMessage(appMessage: AppMessage) {
        Log.d(logTag, "addAppMessage()")
        val currentList = requireNotNull(_appMessageBufferList.value)
        val updateList = currentList.add(appMessage)
        _appMessageBufferList.value = updateList
    }

    fun triggerAppMessageBufferListObserver() {
        Log.d(logTag, "triggerAppMessageBufferListObserver()")
        val currentList = _appMessageBufferList.value
        _appMessageBufferList.value = AppMessageList()
        _appMessageBufferList.value = currentList
    }

    fun removeAppMessageBufferListFirstItem() {
        Log.d(logTag, "removeAppMessageBufferListFirstItem()")
        val currentList = _appMessageBufferList.value
        val updateList = currentList.removeFirstItem()
        _appMessageBufferList.value = updateList
    }

    protected fun equalLastAppMessage(appMessage: AppMessage): Boolean {
        Log.d(logTag, "equalLastAppMessage()")
        val currentList = _appMessageBufferList.value
        return currentList.equalLastItem(appMessage)
    }
}
